clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/CommonServerStoreProxy.java

```diff
@@ -0,0 +1,218 @@
+          LOGGER.debug("CLIENT: doing work to invalidate hash {} from cache {} (ID {})", key, cacheId, invalidationId);
+          for (InvalidationListener listener : invalidationListeners) {
+            listener.onInvalidateHash(key);
+          }
+
+          try {
+            LOGGER.debug("CLIENT: ack'ing invalidation of hash {} from cache {} (ID {})", key, cacheId, invalidationId);
+            entity.invokeAsync(messageFactory.clientInvalidationAck(invalidationId), false);
+          } catch (Exception e) {
+            //TODO: what should be done here?
+            LOGGER.error("error acking client invalidation of hash {} on cache {}", key, cacheId, e);
+          }
+        } else {
+          LOGGER.debug("CLIENT: on cache {}, ignoring invalidation on unrelated cache : {}", messageFactory.getCacheId(), response.getCacheId());
+        }
+      }
+    });
+    this.responseListeners.put(EhcacheEntityResponse.ClientInvalidateAll.class, new EhcacheClientEntity.ResponseListener<EhcacheEntityResponse.ClientInvalidateAll>() {
+      @Override
+      public void onResponse(EhcacheEntityResponse.ClientInvalidateAll response) {
+        final String cacheId = response.getCacheId();
+        final int invalidationId = response.getInvalidationId();
+
+        if (cacheId.equals(messageFactory.getCacheId())) {
+          LOGGER.debug("CLIENT: doing work to invalidate all from cache {} (ID {})", cacheId, invalidationId);
+          for (InvalidationListener listener : invalidationListeners) {
+            listener.onInvalidateAll();
+          }
+
+          try {
+            LOGGER.debug("CLIENT: ack'ing invalidation of all from cache {} (ID {})", cacheId, invalidationId);
+            entity.invokeAsync(messageFactory.clientInvalidationAck(invalidationId), false);
+          } catch (Exception e) {
+            //TODO: what should be done here?
+            LOGGER.error("error acking client invalidation of all on cache {}", cacheId, e);
+          }
+        } else {
+          LOGGER.debug("CLIENT: on cache {}, ignoring invalidation on unrelated cache : {}", messageFactory.getCacheId(), response.getCacheId());
+        }
+      }
+    });
+
+    for (Map.Entry<Class<? extends EhcacheEntityResponse>, EhcacheClientEntity.ResponseListener<? extends EhcacheEntityResponse>> classResponseListenerEntry :
+        this.responseListeners.entrySet()) {
+      this.entity.addResponseListener(classResponseListenerEntry.getKey(), (EhcacheClientEntity.ResponseListener) classResponseListenerEntry.getValue());
+    }
+  }
+
+  @Override
+  public String getCacheId() {
+    return messageFactory.getCacheId();
+  }
+
+  @Override
+  public void addInvalidationListener(InvalidationListener listener) {
+    invalidationListeners.add(listener);
+  }
+
+  @Override
+  public boolean removeInvalidationListener(InvalidationListener listener) {
+    return invalidationListeners.remove(listener);
+  }
+
+  @SuppressWarnings("unchecked")
```

Comment:
This suppress warnings can be removed by introducing a generic type on the method declaration:
`<T extends EhcacheEntityResponse>` which allows the type of both to be bound to one and thus no longer require the cast in line 139